src/Nethermind/Nethermind.Core/Extensions/ByteArrayExtensions.cs

```diff
@@ -56,25 +56,17 @@ public static byte[] Slice(this byte[] bytes, int startIndex, int length)
             return slice;
         }
 
-        public static byte[] SliceWithZeroPaddingEmptyOnError(this byte[] bytes, BigInteger startIndex, int length)
+        public static byte[] SliceWithZeroPaddingEmptyOnError(this byte[] bytes, int startIndex, int length)
         {
-            if (startIndex >= bytes.Length || length == 0)
+            int copiedFragmentLength = Math.Min(bytes.Length - startIndex, length);
+            if (copiedFragmentLength <= 0)
             {
                 return Array.Empty<byte>();
             }
 
-            if (length == 1)
-            {
-                return bytes.Length == 0 ? Array.Empty<byte>() : new[] {bytes[(int)startIndex]};
-            }
-
```

Comment:
wasn't this optimization?

Reply:
Adding here:

    if (length == 1)
    {
        return new[] {bytes[(int)startIndex]};
    }

would definitely make this method faster for a 1-length copy. However, it is making a special case for already quickest-to-process value. If length 1 is used exceptionally frequently - that's cool. Otherwise, it seems useless

Reply:
what if bytes is length of 0?